File: pbark/pbark/config/dog_topics.py
from __future__ import annotations

from pbark.config.config_loader import ConfigLoader
from pbark.config.variable_type import VariableType


class DogTopics:
    FOOD = "food"
    ITEMS = "items"
    INVENTORY = "inventory"
    AGE = "age"
    NAME = "name"
    DESCRIPTION = "description"
    COLOR = "color"
    HAPPY = "happy"
    FED = "fed"

    CANONICAL = [
        FOOD,
        ITEMS,
        INVENTORY,
        AGE,
        NAME,
        DESCRIPTION,
        COLOR,
        HAPPY,
        FED,
    ]

    INT_WORDS = [
        "age",
        "food",
        "items",
        "toys",
        "treats",
        "puppies",
        "beds",
        "squirrels",
        "friends",
        "walks",
        "naps",
        "baths",
        "zoomies",
        "tricks",
    ]

    NAME_WORDS = ["name", "nickname"]
    DESCRIPTION_WORDS = ["description", "collar", "tag", "owner", "family"]
    COLOR_WORDS = ["color", "species", "gender"]
    STRING_WORDS = NAME_WORDS + DESCRIPTION_WORDS + COLOR_WORDS + ["inventory"]
    NICKNAME_WORDS = ["nickname"]
    BOOLEAN_WORDS = [
        "goodboy",
        "goodgirl",
        "gooddog",
        "happy",
        "trained",
        "fed",
        "groomed",
        "washed",
        "vetted",
        "sleepy",
        "tired",
        "excited",
        "muddy",
        "wet",
        "dirty",
        "clean",
        "sick",
        "healthy",
        "rescued",
        "adopted",
        "loyal",
        "stubborn",
        "clever",
        "silly",
        "calm",
        "noisy",
        "quiet",
        "lonely",
    ]

    _FED_WORDS = frozenset({"fed", "groomed", "washed", "vetted"})
    _HAPPY_WORDS = frozenset(
        {
            "goodboy",
            "goodgirl",
            "gooddog",
            "happy",
            "trained",
            "sleepy",
            "tired",
            "excited",
            "muddy",
            "wet",
            "dirty",
            "clean",
            "sick",
            "healthy",
            "rescued",
            "adopted",
            "loyal",
            "stubborn",
            "clever",
            "silly",
            "calm",
            "noisy",
            "quiet",
            "lonely",
        }
    )
    _FOOD_WORDS = frozenset(
        {
            "food",
            "treat",
            "treats",
            "snack",
            "snacks",
            "cookie",
            "cookies",
            "biscuit",
            "biscuits",
        }
    )
    _ITEM_WORDS = frozenset({"item", "items", "toy", "toys"})
    _CANONICAL_SET = frozenset(CANONICAL)
    _ALIAS = None

    @classmethod
    def _aliases(cls) -> dict[str, str]:
        if cls._ALIAS is None:
            cls._ALIAS = cls._build_aliases()
        return cls._ALIAS

    @staticmethod
    def default_assign_topic() -> str:
        return DogTopics.AGE

    @staticmethod
    def default_adjust_topic() -> str:
        return DogTopics.ITEMS

    @staticmethod
    def default_food_adjust_topic() -> str:
        return DogTopics.FOOD

    @classmethod
    def resolve(cls, word: str | None) -> str | None:
        if word is None:
            return None
        normalized = ConfigLoader.normalise(word)
        mapped = cls._aliases().get(normalized)
        if mapped is not None:
            return mapped
        if ConfigLoader.type_of(normalized) == VariableType.OBJECT:
            return None
        if normalized in cls._CANONICAL_SET:
            return normalized
        return None

    @classmethod
    def is_field_word(cls, word: str) -> bool:
        return cls.resolve(word) is not None

    @staticmethod
    def is_food_field(field: str) -> bool:
        return DogTopics.FOOD == field

    @staticmethod
    def is_items_field(field: str) -> bool:
        return DogTopics.ITEMS == field

    @staticmethod
    def is_numeric_field(field: str) -> bool:
        return field in (DogTopics.FOOD, DogTopics.ITEMS, DogTopics.AGE)

    @classmethod
    def _build_aliases(cls) -> dict[str, str]:
        m: dict[str, str] = {}
        for word in cls._FOOD_WORDS:
            m[word] = cls.FOOD
        for word in cls._ITEM_WORDS:
            m[word] = cls.ITEMS
        for extra in ("puppy", "bed", "walk", "nap", "trick", "friend", "squirrel"):
            m[extra] = cls.ITEMS
        for word in cls.INT_WORDS:
            if word not in (cls.AGE, cls.FOOD, cls.ITEMS):
                m.setdefault(word, cls.ITEMS)
        for word in cls.NAME_WORDS:
            m.setdefault(word, cls.NAME)
        for word in cls.DESCRIPTION_WORDS:
            m.setdefault(word, cls.DESCRIPTION)
        for word in cls.COLOR_WORDS:
            if word != cls.COLOR:
                m.setdefault(word, cls.COLOR)
        for word in cls.BOOLEAN_WORDS:
            if word in cls._FED_WORDS:
                m.setdefault(word, cls.FED)
            elif word in cls._HAPPY_WORDS:
                m.setdefault(word, cls.HAPPY)
        m["inventory"] = cls.INVENTORY
        m["belongings"] = cls.INVENTORY
        return dict(m)

File: src/main/java/dev/klomptech/jbark/config/DogTopics.java
package dev.klomptech.jbark.config;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

// Dog fields: food and items are separate counters, plus inventory for held objects
// Story words like toys or treats map to the right counter
public final class DogTopics {

  // Treats, snacks, stolen cookies. Greedy applies here.
  public static final String FOOD = "food";
  // Toy count on the collar. Shares and snags use this.
  public static final String ITEMS = "items";
  // Objects the dog is carrying (ball, stick, etc.).
  public static final String INVENTORY = "inventory";

  public static final String AGE = "age";
  public static final String NAME = "name";
  public static final String DESCRIPTION = "description";
  public static final String COLOR = "color";
  public static final String HAPPY = "happy";
  public static final String FED = "fed";

  public static final List<String> CANONICAL =
      List.of(FOOD, ITEMS, INVENTORY, AGE, NAME, DESCRIPTION, COLOR, HAPPY, FED);

  public static final List<String> INT_WORDS =
      List.of(
          "age",
          "food",
          "items",
          "toys",
          "treats",
          "puppies",
          "beds",
          "squirrels",
          "friends",
          "walks",
          "naps",
          "baths",
          "zoomies",
          "tricks");

  public static final List<String> NAME_WORDS = List.of("name", "nickname");

  public static final List<String> DESCRIPTION_WORDS =
      List.of("description", "collar", "tag", "owner", "family");

  public static final List<String> COLOR_WORDS = List.of("color", "species", "gender");

  public static final List<String> STRING_WORDS =
      List.of(
          "name",
          "nickname",
          "description",
          "collar",
          "tag",
          "owner",
          "family",
          "color",
          "species",
          "gender",
          "inventory");

  public static final List<String> NICKNAME_WORDS = List.of("nickname");

  public static final List<String> BOOLEAN_WORDS =
      List.of(
          "goodboy",
          "goodgirl",
          "gooddog",
          "happy",
          "trained",
          "fed",
          "groomed",
          "washed",
          "vetted",
          "sleepy",
          "tired",
          "excited",
          "muddy",
          "wet",
          "dirty",
          "clean",
          "sick",
          "healthy",
          "rescued",
          "adopted",
          "loyal",
          "stubborn",
          "clever",
          "silly",
          "calm",
          "noisy",
          "quiet",
          "lonely");

  private static final Set<String> FED_WORDS = Set.of("fed", "groomed", "washed", "vetted");

  private static final Set<String> HAPPY_WORDS =
      Set.of(
          "goodboy",
          "goodgirl",
          "gooddog",
          "happy",
          "trained",
          "sleepy",
          "tired",
          "excited",
          "muddy",
          "wet",
          "dirty",
          "clean",
          "sick",
          "healthy",
          "rescued",
          "adopted",
          "loyal",
          "stubborn",
          "clever",
          "silly",
          "calm",
          "noisy",
          "quiet",
          "lonely");

  private static final Set<String> FOOD_WORDS =
      Set.of(
          "food", "treat", "treats", "snack", "snacks", "cookie", "cookies", "biscuit", "biscuits");

  private static final Set<String> ITEM_WORDS = Set.of("item", "items", "toy", "toys");

  private static final Map<String, String> ALIAS = buildAliases();

  private DogTopics() {}

  public static String defaultAssignTopic() {
    return AGE;
  }

  // Lines like "she pinches a treat" use the treat word. Plain "she finds a toy" adds to items
  public static String defaultAdjustTopic() {
    return ITEMS;
  }

  public static String defaultFoodAdjustTopic() {
    return FOOD;
  }

  public static String resolve(final String word) {
    if (word == null) {
      return null;
    }
    String normalized = ConfigLoader.normalise(word);
    String mapped = ALIAS.get(normalized);
    if (mapped != null) {
      return mapped;
    }
    if (ConfigLoader.typeOf(normalized).orElse(null) == VariableType.OBJECT) {
      return null;
    }
    if (CANONICAL_SET.contains(normalized)) {
      return normalized;
    }
    return null;
  }

  public static boolean isFieldWord(final String word) {
    return resolve(word) != null;
  }

  public static boolean isFoodField(final String field) {
    return FOOD.equals(field);
  }

  public static boolean isItemsField(final String field) {
    return ITEMS.equals(field);
  }

  public static boolean isNumericField(final String field) {
    return FOOD.equals(field) || ITEMS.equals(field) || AGE.equals(field);
  }

  private static final Set<String> CANONICAL_SET = Set.copyOf(CANONICAL);

  private static Map<String, String> buildAliases() {
    Map<String, String> map = new HashMap<>();
    for (String word : FOOD_WORDS) {
      map.put(word, FOOD);
    }
    for (String word : ITEM_WORDS) {
      map.put(word, ITEMS);
    }
    map.put("puppy", ITEMS);
    map.put("bed", ITEMS);
    map.put("walk", ITEMS);
    map.put("nap", ITEMS);
    map.put("trick", ITEMS);
    map.put("friend", ITEMS);
    map.put("squirrel", ITEMS);
    for (String word : INT_WORDS) {
      if (!AGE.equals(word) && !FOOD.equals(word) && !ITEMS.equals(word)) {
        map.putIfAbsent(word, ITEMS);
      }
    }
    for (String word : NAME_WORDS) {
      map.putIfAbsent(word, NAME);
    }
    for (String word : DESCRIPTION_WORDS) {
      map.putIfAbsent(word, DESCRIPTION);
    }
    for (String word : COLOR_WORDS) {
      if (!COLOR.equals(word)) {
        map.putIfAbsent(word, COLOR);
      }
    }
    for (String word : BOOLEAN_WORDS) {
      if (FED_WORDS.contains(word)) {
        map.putIfAbsent(word, FED);
      } else if (HAPPY_WORDS.contains(word)) {
        map.putIfAbsent(word, HAPPY);
      }
    }
    map.put("inventory", INVENTORY);
    map.put("belongings", INVENTORY);
    return Map.copyOf(map);
  }
}
